NDTensors/src/default_storage.jl

```diff
@@ -0,0 +1,19 @@
+## This is a fil which specifies the default storage type provided some set of parameters
+## The parameters are the element type and storage type
+default_datatype(eltype::Type{<:Number}) = Vector{eltype}
+default_eltype() = Float64
+
+default_storagetype(datatype::Type{<:AbstractArray{ElT}}) where {ElT} = Dense{ElT, datatype}
+
+default_storagetype(datatype::Type{<:AbstractArray}) = default_storagetype(datatype{default_eltype()})
```

Comment:
Wouldn't this overwrite the version above, and also overwrite an existing element type?

Reply:
No so these are actually 2 different cases. The first one covers if you provide a datatype with a elementtype hard coded or deduced from previous lines of code. The second one covers if you just provide a datatype without any type information. By default `Vector` is a `Array{T, 1} where {T}` so `eltype(Vector) = Any`. To mitigate problems of multiple types in a vector I require a call with an untyped datatype to use the default element type

Reply:
Indeed you're correct, I didn't appreciate the difference in how Julia dispatches one those two signatures:
```julia
julia> f(T::Type{<:AbstractArray{ElT}}) where {ElT} = "f1"
f (generic function with 1 method)

julia> f(T::Type{<:AbstractArray})= "f2"
f (generic function with 2 methods)

julia> f(Vector{Int})
"f1"

julia> f(Vector)
"f2"
```
That's a nice way to differentiate whether or not an element type is already specified!

I think instead of:
```julia
default_storagetype(datatype::Type{<:AbstractArray}) = default_storagetype(datatype{default_eltype()})
```
we should do something like:
```julia
default_storagetype(datatype::Type{<:AbstractArray}) = default_storagetype(adapt_type(default_eltype(), datatype))
```
(here I'm using `adapt_type` which is the next generation of the current function `similartype`: https://github.com/ITensor/ITensors.jl/blob/v0.3.23/NDTensors/src/similar.jl). Otherwise it wouldn't preserve other type parameters like the dimensions of the AbstractArray, or handle array wrapper correctly.

Reply:
Is adapt_type currently functional/implemented?

Reply:
Unfortunately no, I was just talking to Miles and he needs the same kind of functionality so we should meet about implementing it. It could even get turned into a general Julia package `AdaptTypes.jl`.

But for the sake of this PR for now you may be able to use `similartype`.

Reply:
Okay, I know `similartype` does not work with CuVectors currently but I will still include it on this PR